// @stonksjs/finviz/dist/cjs/Finviz.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const cheerio_1 = require("cheerio");
const camelCase_1 = __importDefault(require("lodash/camelCase"));
const helpers_1 = require("./helpers");
const { FINVIZ_URL, SCREENER_URL, DEFAULT_QUOTE } = helpers_1.contsants;
class Finviz {
    constructor() {
        this.api = axios_1.default.create({
            baseURL: FINVIZ_URL,
        });
        this.quoteCache = new Map();
        this.screenerCache = new Map();
        this.screenrerListCache = new Set();
    }
    /**
     * Parses a given HTML string from Finviz and returns a Quote object
     * @param {string} html - html string
     * @returns {Quote} - quote object
     */
    static parseQuoteData(html) {
        const $ = (0, cheerio_1.load)(html);
        const rows = $('.snapshot-table2 > tbody > tr');
        const data = {};
        let lastKey = '';
        rows.each((_, row) => {
            const cells = $(row).find('> td');
            cells.each((j, cell) => {
                const value = $(cell).text();
                if (j % 2 === 0) {
                    lastKey = (0, camelCase_1.default)(value);
                }
                else {
                    const cleanValue = value.replace(/,/g, '');
                    const parsedValue = (0, helpers_1.isNumeric)(cleanValue)
                        ? parseFloat(cleanValue)
                        : value;
                    // @ts-ignore
                    data[lastKey] = parsedValue;
                }
            });
        });
        return Object.assign(Object.assign({}, DEFAULT_QUOTE), data);
    }
    /**
     * Get quote data for a given symbol
     * @param {string} symbol - stock symbo, e.g. AAPL
     * @returns {Quote} - quote object
     */
    getQuote(symbol) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.quoteCache.has(symbol)) {
                return this.quoteCache.get(symbol);
            }
            const url = `/quote.ashx?t=${symbol}`;
            const response = yield this.api.get(url);
            const data = Finviz.parseQuoteData(response.data);
            this.quoteCache.set(symbol, data);
            return data;
        });
    }
    /**
     *
     * Parses a given HTML string from Finviz and returns an array of symbols
     * @param {string} html - html string
     * @returns {string[]} - array of stock symbols
     */
    static parseScreenerResponse(html) {
        const $ = (0, cheerio_1.load)(html);
        const rows = $('.table-light > tbody > tr:not(:first-child)');
        const output = [];
        rows.each((_, row) => {
            const cell = $(row).find('> td:nth-child(2)');
            if (cell.length) {
                output.push(cell.text());
            }
        });
        return output;
    }
    /**
     * Gets a list of pre-defined screeners from Finviz
     * @returns {Screener[]} - array of available screeners
     */
    getScreenersList() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.screenrerListCache.size) {
                return [...this.screenrerListCache];
            }
            const { data: html } = yield this.api.get(SCREENER_URL);
            const $ = (0, cheerio_1.load)(html);
            const options = $('#signalSelect > option');
            const output = [];
            options.each((_, option) => {
                const { value } = option.attribs;
                if (!value.includes('&s='))
                    return;
                const name = option.children[0].data;
                const [, query] = value.split('&s=');
                const url = new URL(value, SCREENER_URL).href;
                const id = query.split('_').pop();
                output.push({ name, url, id });
            });
            this.screenrerListCache = new Set(output);
            return output;
        });
    }
    /**
     * Returns a url for a given Finviz screener
     * @param {string} screener - Finviz screener name
     * @returns {string} - screener url
     */
    getUrlForScreener(screener) {
        return __awaiter(this, void 0, void 0, function* () {
            const screeners = yield this.getScreenersList();
            const output = screeners.find((s) => s.name === screener || s.id === screener);
            if (!output) {
                throw new Error(`Screener ${screener} not found`);
            }
            return output.url;
        });
    }
    /**
     * Given a Finviz screener url, returns an array of symbols
     * @param {string} screener - Finviz screener name or url
     * @returns {string[]} - array of stock symbols
     */
    getScreener(screener) {
        return __awaiter(this, void 0, void 0, function* () {
            let url = screener;
            if (!(0, helpers_1.isUrl)(url)) {
                url = yield this.getUrlForScreener(screener);
            }
            if (this.screenerCache.has(url)) {
                return this.screenerCache.get(url);
            }
            const { data } = yield this.api.get(url);
            const results = Finviz.parseScreenerResponse(data);
            this.screenerCache.set(url, results);
            return results;
        });
    }
}
exports.default = Finviz;
